### PR TITLE
Underscores

### LINKED ISSUE #14: the real world isn't RFC compliant
Unfortunately the real world does not listen to RFCs, `_` can be in hostname parameters, DNS respects it, chrome respects it but we fail here.

We should introduce a `strict` mode and a `loose` mode for trying to allow for better real world compatibility where required.

### PR DESCRIPTION
**Pull request by ypcrts**
#15 merged with develop
resolves #14 

supports underscores, optionally

### CODE REVIEW

File: tests/test_fqdn.py [hunk 1]
```diff
@@ -154,82 +179,111 @@ def test_min_labels_valid_set_to_3(self):
...
+        with pytest.raises(ValueError):
+            FQDN("trainwreckcom", allow_underscores=a_u).absolute
+
+    def test_relative_fqdn_true(self, a_u):
+        assert FQDN("relative.com", allow_underscores=a_u).is_valid_relative is True
 
-    def test_relative_fqdn_true(self):
-        assert FQDN("relative.com").is_valid_relative is True
+    def test_relative_fqdn_false(self, a_u):
+        assert FQDN("relative.com.", allow_underscores=a_u).is_valid_relative is False
```
Comment (ypcrts) [inline tests/test_fqdn.py:208]: 💩

**Comment by wakemaster39** [comment]
This looks good to me, its weird allowing the domain name to have underscores in it even though the CA's block it but I suppose if chromium is allowing it, it makes sense. It also feels weird with an underscore leading a domain mostly because those are identifiers for other DNS entry types like DKIM but again if chromium allows it, it makes sense to do the same.

**Comment by wakemaster39** [comment]
The only other thing I would say about a different change you made in develop is making flake8 part of pre-commit.

Love pre-commit and had it setup locally so I was making sure to black things, but I always hate flake8 as part of it. I personally allow code to be committed poorly, so that I can quickly expose PRs for review and the like, and then include it as part of the build before it can be merged. 

Just a preference thing, but black 100% part of pre-commit.

**Comment by wakemaster39** [comment]
Sorry another question about dev, why did `    install_requires=["cached-property>=1.3.0;python_version<'3.8'"],` get rolled back to not having the version string?

**Comment by ypcrts** [comment]
this is a fail on my part. a setuptools linter in my editor yelled at me. oops

**Comment by ypcrts** [comment]
yeah, i agree with you flake8 is pain. most of the formatting errors should be non-blocking; i should configure this correctly. i haven’t been a full-time dev in a few years now so i’m probably slightly out of date on exactly how to do this best.

black doesn’t do linting though. flake8 pointed out unused imports and use before assignment errors.

i’ll look into this deeper

**Pull request merged**